execute/row_fn.go

```diff
@@ -13,52 +12,85 @@ import (
 )
 
 type dynamicFn struct {
-	compilationCache *compiler.CompilationCache
-	inRecord         values.Object
-
-	preparedFn compiler.Func
-
+	// Configuration attributes. These are initialized once
+	// on creation and used for each new compilation.
+	scope      compiler.Scope
+	fn         *semantic.FunctionExpression
 	recordName string
-	record     *Record
 
-	recordCols map[string]int
-	references []string
+	// These are initialized after the function is prepared.
+	preparedFn compiler.Func
+	arg0       values.Object
+	args       values.Object
 }
 
 func newDynamicFn(fn *semantic.FunctionExpression, scope compiler.Scope) dynamicFn {
 	return dynamicFn{
-		compilationCache: compiler.NewCompilationCache(fn, scope),
-		inRecord:         values.NewObject(),
-		recordName:       fn.Block.Parameters.List[0].Key.Name,
-		references:       findColReferences(fn),
-		recordCols:       make(map[string]int),
+		scope:      scope,
+		fn:         fn,
+		recordName: fn.Block.Parameters.List[0].Key.Name,
+	}
+}
+
+// typeof returns an object monotype that matches the current column data.
+func (f *dynamicFn) typeof(cols []flux.ColMeta) (semantic.MonoType, error) {
+	properties := make([]semantic.PropertyType, len(cols))
+	for i, c := range cols {
+		vtype := flux.SemanticType(c.Type)
+		if vtype.Kind() == semantic.Unknown {
+			return semantic.MonoType{}, errors.Newf(codes.Internal, "unknown column type: %s", c.Type)
+		}
+		properties[i] = semantic.PropertyType{
+			Key:   []byte(c.Label),
+			Value: vtype,
+		}
 	}
+	return semantic.NewObjectType(properties), nil
 }
 
 func (f *dynamicFn) prepare(cols []flux.ColMeta, extraTypes map[string]semantic.MonoType) error {
```

Comment:
Is extraTypes still necessary?

Reply:
I think so. I think it's needed for reduce but I haven't gotten there yet honestly.